src/toil/batchSystems/mesos/executor.py

```diff
@@ -94,6 +103,42 @@ def _sendStats(self, driver):
             log.debug("Sent stats message")
             sleep(30)
 
+    def _sendCurlRequest(self, endpoint, slaveID='', delta=''):
+        """
+        Fills in the required bits of a GET/POST message to a mesos master HTTP endpoint (slaves,
+        reserve, or unreserve) and sends the request.
+        :param str endpoint: Endpoint to send the request to. ('slaves', 'reserve', or 'unreserve')
+        :param str slaveID: Id of the slave. required for reserve and unreserve requests
+        :param int delta: The value to reserve/unreserve.
+        :return: urllib.urlopen construct of the sent request
+        """
+        assert endpoint in ['slaves', 'reserve', 'unreserve']
+        # For a reserve/unreserve operation, you need a delta and a slave to operate on.
+        if endpoint != 'slaves':
+            assert delta
+            assert slaveID
+        url = 'http://' + self.masterAddress + '/master/' + endpoint
+        with open(self.mesosCredentials, 'r') as cF:
+            username, password = cF.readline().strip().split()
+        passMan = urllib2.HTTPPasswordMgrWithDefaultRealm()
+        passMan.add_password(None, url, username, password)
+        auth_handler = urllib2.HTTPBasicAuthHandler(passMan)
+        opener = urllib2.build_opener(auth_handler)
+        urllib2.install_opener(opener)
+        data = None
+        if delta:
+            data = urllib.urlencode({'slaveId': slaveID,
+                                     'resources': json.dumps([{"name": "disk",
+                                                               "type": "SCALAR",
+                                                               "scalar": { "value": delta},
+                                                               "role": "toil-leftover",
+                                                               "reservation": {"principal": "toil"}}
+                                                              ])})
+        log.debug("Sending a request to the mesos master endpoint %s" % endpoint)
+        # separating this onto 2 line so the stack trace is cleaner
+        content = urllib2.urlopen(url, data)
```

Comment:
The deallocation happens if the endpoint is specified at unreserve. The only difference between a reserve and unreserve operation is the endpoint. Everything else remains constant.

If the finction received `endpoint` was passed as `reserve` or `unreserve`,  then we assert that a `delta` and the `slaveId` have also been passed. The delta is the amount of disk to be reserved or unreserved. I have a  test that checks the behaviour of both (reserve and unreserve).